### DISCUSSION TITLE
Help Interpreting: Using “tokens” for run-time variable replacement in shows

### DISCUSSION THREAD

**Discussion by ttrickel** [missionpinball/missionpinball#79] [Q&A]
On [this](https://docs.missionpinball.org/en/stable/shows/tokens.html) page includes the following shows.

##! show: my_show
- time: 0
  lights:
    (led): red
- time: 1
  lights:
    (led): off

show_player:
  some_event:
    flash_red:
      loops: -1
      show_tokens:
        led: led_02

My question is how to get this to work in a config file.

Should the first snippet be:

shows:
  flash_red:
    -time: 0
    lights:
      (led): off
    - time: 1
    lights:
      (led): off

If so, I must be doing something else wrong as I get an "ValueError: YAML error found in file C:\Users\student\pinball\cobraPinTest1\modes\reentry\config\reentry.yaml. Line 84, Position 16: mapping values are not allowed here"

Do I have to put the flash_red show in its own folder structure?

**Comment by atummons** [comment]
Two things. Your version shows off in both steps, so it'll never do anything.

Second, can you post and highlight and use the code option at the top. Looks similar to <>. This will retain indents. I would guess that the issue that something isn't indented correctly.

**Reply by ttrickel** [reply]
Thanks atummons, Let me try that again.

Here's the code from [this page](https://docs.missionpinball.org/en/stable/shows/tokens.html): (took a bit to understand how the code  option works)

```
##! show: my_show
- time: 0
  lights:
    (led): red
- time: 1
  lights:
    (led): off
```
...
```
show_player:
  some_event:
    flash_red:
      loops: -1
      show_tokens:
        led: led_02
```

I see where my code had off in both steps. But that shouldn't generate a ValueError should it?

Perhaps what I'm asking is how does the second snippet work with the first snippet?

Should the first snippet be in a flash_red config file and a show (not mode) directory named flash_red? Then why not ##! show: flash_red

**Reply by atummons** [reply]
There are multiple different ways to do shows. You can do them in your mode config, or you can do it in a separate show. The show would need to be in a show folder. i.e.,mode/shows/myshow structure.

The difference is the first snippet is "creating" the show. Meaning I have a set of actions that can be performed whenever I want. When you do, it'll go red/off alternating.

The second is calling the show, it says when the event happens, I want you to play my show. You can tell it a lot of things, duration, loops, speed, etc.

**Reply by ttrickel** [reply]
Thanks atummons,

You're doing a great job with the documentation. There's just so many ways to do stuff - which is good. It's harder for me to get my head around it.

I'll leave this thread with the question - ##! show: my_show is short hand for this should be in a file named my_show correct? If so, then where would flash_red be? Could it be ##! show: flash_red?

Thanks again
